### ISSUE TITLE
When asking for a component with an unspecified version (i.e. 1.x.x) the response should not contain custom headers

### ISSUE THREAD

**Issue by ivan-dimitrov-skyscanner** [opentable/oc#333]
This issue is especially valid when the custom headers are cache related. In this case if the client asks for hello-world 1.x.x and it returns cache headers then the intermediate HTTP proxy agents may cache the result, which will be wrong.
My suggestion in this case it to not return custom headers.
@matteofigus, what are your thoughts on that?

**Comment by matteofigus** [comment]
I am not sure this is an issue but an actually intended behaviour.

When you say "the intermediate HTTP proxy agents may cache the result, which will be wrong" - I actually don't see why that would be wrong. The response contains both the resolved version and the requested version for the given component. You can build the proxy agent's cache using the resolved version as key, and in parallel implement a polling mechanism (I think this is the main goal for #335) to establish when to clear that cache (because you would have a new version, so you need to make sure, for instance, that new reqs for 1.2.x forget about 1.2.45 and its cache and start serving 1.2.46).

Perhaps I am missing something? Maybe you can include here an example?

**Comment by ivan-dimitrov-skyscanner** [comment]
Hi @matteofigus, 

In case the proxy is the one constructed especially for caching components then sure, it can be made in a way to accept this response and work with it.

However in this issue I'm speaking more of the standard HTTP proxies like varnish or nginx.
For example when you send the request:
http://registry_url/hello-world/1.0.1
and the registry responds adding a custom cache header like - Cache-Control: public max-age=86400 we are OK with the proxy to cache this result for 24 hours because the component's designer expects that the component will not change the content in the next XX hours.

On the other hand if we send a request with unspecified version like:
http://registry_url/hello-world/1.x.x and the registry responds with 
hello-world version 1.0.1 and again the cache header - Cache-Control: public max-age=86400 the proxy will again cache the result but in this case it will correspond to version 1.x.x. If then, for whatever reason, a new version of this component is introduces meanwhile (i.e. 1.1.0 or 1.0.2) every subsequent request made for 1.x.x will still return 1.0.1 for some time.

**Comment by matteofigus** [comment]
Good point - still I guess it's part of the game for when doing caching in the component's level.
I think I would prefer to add this extra logic to the components for cleanness and enhanced flexibility. For instance, it may deliver caching to a subset of consumers - or even deliver different cache values. I am quite sure in case of nginx or varnish we can observe the request headers and act differently.

Example:
```js
module.exports.data = function(context, callback){
  if(context.requestHeaders('user-agent') === 'Varnish'){
    context.setHeader('Cache-Control', 'nope! 💩');
  }
  ...
};
```

Another thing, a bit unrelated I guess, you mention 24 hours for caching - which may be good for you, but I guess my vision is to always have 5 to 20 minutes caches set by components. That would be in my opinion enough for you to drastically reduce load to your servers (which I guess is the main reason for using caching) but to have still a very decently predictable time for when you want a change to propagate for whatever reason - but that's just my 2 cents 👍

**Comment by ivan-dimitrov-skyscanner** [comment]
Completely agree with you about the specific headers based on the different "surrounding environment".
What we are aiming to achieve is to be able to cache the response literally forever for 90% of the component requests. The idea is to separate the very dynamic content and deliver it through AJAX calls while the component will be rendered based on its name, version and the specific parameters set in the request. This should render content that should stay persistent for very long time.

**Comment by mh52** [comment]
Hi @matteofigus, I'm not sure the code snippet above will be sufficient for our use-case - we only don't want to cache the component response when the request to the registry isn't for a specific version e.g. `1.x.x`. Whereas if the request was for a specific version e.g. `1.0.1` then we would want to cache.

I don't expect the component itself will have awareness of the incoming request to the registry, and so won't be able to determine whether a cache-header value should be set or not?

E.g. I expect it would need to be something like (pseudocode-ish):

```
module.exports.data = function(context, callback){
  if(request.version.contains('x')){ // e.g. 1.x.x
    context.setHeader('Cache-Control', 'nope! 💩');
  } else { // e.g. 1.0.1
    context.setHeader('Cache-Control', 'max-age=<seconds>');
}
  ...
};
```

This logic will apply across all components for our use-cases and would prefer it to exist in a central place, rather than relying on many individual component owners to implement the same logic.

We expect a high cache rate for a large majority of our components, while still being able to benefit from the flexibility of using the OC registry to manage version releases, which aren't impacted by long cache times (as Ivan has explained).

It also doesn't feel ideal for the registry to be overriding specific behaviour (i.e. sometimes omitting headers set by the components in this example), so would be good to investigate other possible approaches to this.

- Do components have visibility of whether the incoming request is for a specific version or 'x.x.x' version? If so, that would make the above code snippet one potential option to investigate further (but with the concerns mentioned above).

- Is there a more explicit way of doing this? e.g. when the component sets any headers, it specifies whether the header should be applied for exact version requests only or not? This setting can then be obeyed by the registry when processing incoming component requests.

Just throwing some suggestions into the mix :)

**Comment by matteofigus** [comment]
(apologies for the delay - on holiday now)

Very good ideas @mh52 

* I think implementing the first (exposing the requested version in the context object) is a good feature to add anyways and I think it may not require so much work.
* In regards of the second (bypassing the response headers in case of version request) - I think I still am not 100% sure, but in case we proceed I would make it an optional setting on the registry configuration - something like `bypassResponseHeadersOnWeakVersionMatching=true|false` (with a better name maybe 😄) so that users could override this only if needed (with a default to false). Or perhaps a `blacklistedResponseHeadersOnWeakVersionMatching=[array]` with a default empty in case we need extra powers.

Another quick note for centralising logic: you can actually write plugins for centralising logic shared between components. We do that all the time. For instance, you could have something like that in each component:

```js
module.exports.data = function(context, callback){
  context.plugins.setCachePolicies(context);
  ...
};
```

and then create your plugin in the registry initialisation:

```js
registry.register({
  name: 'setCachePolicies',
  register: {
    register: function(options, dependencies, next){
      // useful if you have some async initialisation phase
      next();
    },
    execute: function(ctx){
      // the functionality you want to expose
      if(ctx.requestVersion.contains('x')){ ctx.setHeader(...); }
      ...
    }
  }
});
```
A plugin example: https://github.com/opentable/oc-statsd

**Comment by ivan-dimitrov-skyscanner** [comment]
Hi @matteofigus,

Having plugins to control specific behaviours is appealing and fit well in the general picture. However this will work well only when introducing the behaviour/feature with a plugin from the very beginning. If we introduce it later (as in this case) it will require changes in every component in order to re-configure the behaviour. This is especially valid if we have a big component inventory with third party components.

Thus I like the second approach with configuring the registry to behave in a specific way when returning custom headers. If I have to pick up one solution perhaps it would be the one with: "blacklistedResponseHeadersOnWeakVersionMatching=[array] with a default empty in case we need extra powers" but let us think a little be more on that, too.

**Comment by mh52** [comment]
@matteofigus thanks for your additional ideas and plugin example 👍 

@ivan-dimitrov-skyscanner and I discussed this a bit more today and would like to try out the following (based off your idea):

- Introduce a registry config `responseHeadersToOmitOnWeakVersion = []`.
- Use this new config within the registry to omit certain headers from the response when a weak version request is received.
- We'll populate this list with `Cache-Control` for our specific use-case; it'll be empty by default.

Reasons for deciding on this approach:

- It provides consistent behaviour across all components, which is important for our use-case.
- It avoids having dependencies and expectations on multiple teams to implement and own this logic within each individual component (which we expect to have a lot of).
- Restricting the behaviour to only the headers specified in `responseHeadersToOmitOnWeakVersion` avoids unnecessarily wider impact i.e. by only removing specified headers from responses for weak-versioned requests, rather than blanket behaviour to needlessly remove all headers.

Let us know if you've got any other suggestions/concerns with this approach - we'll prepare a pull request with the proposed update. Thanks

**Comment by matteofigus** [comment]
Sounds great to me. Thanks for your help 👍

**Comment by ivan-dimitrov-skyscanner** [comment]
I'll start working in this change ;-)
I am still not sure how to name the configuration parameter. What I'm thinking of is among the following options:
- headersToSkipOnWeakVersion
- customHeadersToSkipOnWeakVersion (perhaps I would go with this)
- headersToOmitOnWeakVersion
- customHeadersToOmitOnWeakVersion

@matteofigus, @mh52, what are your preferences?

**Comment by matteofigus** [comment]
+1 on `customHeadersToSkipOnWeakVersion`

**Comment by matteofigus** [comment]
Fixed by #343
